<?xml version="1.0" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<title>MAKE_IMAGES_HOWTO</title>
<meta http-equiv="content-type" content="text/html; charset=utf-8" />
<link rev="made" href="mailto:root@localhost" />
</head>

<body style="background-color: white">


<!-- INDEX BEGIN -->
<div name="index">
<p><a name="__index__"></a></p>

<ul>

	<li><a href="#gbrowse_img___cgi_script_to_generate_genome_images_via_the_generic_genome_browser">gbrowse_img - CGI script to generate genome images via the Generic Genome Browser</a></li>
	<li><a href="#synopsis">SYNOPSIS</a></li>
	<li><a href="#description">DESCRIPTION</a></li>
	<li><a href="#cgi_arguments">CGI arguments</a></li>
	<li><a href="#known_bugs">KNOWN BUGS</a></li>
	<li><a href="#support">SUPPORT</a></li>
	<li><a href="#author">AUTHOR</a></li>
</ul>

<hr name="index" />
</div>
<!-- INDEX END -->

<p>
</p>
<h1><a name="gbrowse_img___cgi_script_to_generate_genome_images_via_the_generic_genome_browser">gbrowse_img - CGI script to generate genome images via the Generic Genome Browser</a></h1>
<p>
</p>
<hr />
<h1><a name="synopsis">SYNOPSIS</a></h1>
<pre>
  &lt;img src=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3;width=400">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3;width=400</a>&quot;&gt;</pre>
<p>Will generate this picture:</p>
<img src="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3;width=400"><br><pre>
   &lt;a href=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img?list=sources">http://www.wormbase.org/db/seq/gbrowse_img?list=sources</a>&quot;&gt;list&lt;/a&gt;</pre>
<p>Will return this document:</p>
<pre>
  ## Sources
  human
  ws77
  worm_pmap
  embl
  fly
  briggsae
  yeast
  wormbase</pre>
<pre>
   &lt;a href=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img?list=types">http://www.wormbase.org/db/seq/gbrowse_img?list=types</a>&quot;&gt;types&lt;/a&gt;</pre>
<p>Will return this document:</p>
<pre>
        ## Feature types for source yeast
        Genes   Named gene      default
        ORFs    ORF     default
        CDS     CDS     
        tRNAs   tRNAs   default
        Centro  Centromeres     
        Transp  Transposons     
        LTRs    Long Terminal Repeats   
        TranslationF    3-frame translation (forward)   
        DNA/GC Content          
        TranslationR    3-frame translation (reverse)   
        ncRNA   Noncoding RNAs</pre>
<p>
</p>
<hr />
<h1><a name="description">DESCRIPTION</a></h1>
<p>This CGI script is an interface to the Generic Genome Browser for the
purpose of retrieving dynamic images of a region of the genome. It can
be used as the destination of an &lt;img&gt; tag like this:</p>
<pre>
  &lt;img src=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=III:1">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=III:1</a> ..1000&quot;&gt;</pre>
<p>The script can also be used to superimpose one or more external
features onto the display, for example for the purpose of displaying
BLAST hits, an STS or a knockout in the context of the genome.</p>
<p>
</p>
<hr />
<h1><a name="cgi_arguments">CGI arguments</a></h1>
<p>The script recognizes the following CGI arguments, which can be passed
either as GET or POST argument=value pairs. Argument pairs must be
separated by semicolons (preferred) or by ampersands.</p>
<pre>
   Argument     Alias   Description
   name         q       genomic landmark or range
   type         t       tracks to include in image
   width        w       desired width of image
   options      o       list of track options (compact, labeled, etc)
   abs          b       display position in absolute coordinates
   add          a       added feature(s) to superimpose on the image
   style        s       stylesheet for additional features
   keystyle     k       where to place the image key
   overview             force an overview-style display
   flip         f       flip image left to right
   embed                generate full HTML for image and imagemap for use in an
                        embedded frame
   format               format  for  the  image  (use  &quot;SVG&quot;  for  scaleable vector
                        graphics)
   list                 get certain types of configuration information
   source               database name</pre>
<p>The arguments are explained in more detail here</p>
<dl>
<dt><strong><a name="name" class="item">name (Alias: q)</a></strong></dt>

<dd>
<p>This argument specifies the region of the genome to be
displayed. Several forms are recognized:</p>
<p>name=Landmark</p>
<p>Display the landmark named &quot;Landmark&quot;. Valid landmark names include
chromosomes, contigs, clones, STSs, predicted genes, and any other
landmark that the administrator has designated. Be careful when
fetching large landmarks such as whole chromosomes!</p>
<p>name=Landmark:start..end</p>
<p>Display the region between start and end relative to &quot;Landmark&quot;.</p>
<p>name=Class:Landmark</p>
<p>Display &quot;Landmark&quot;, restricting to a particular class, such as
&quot;PCR_Product&quot;. The list of classes is under the control of the
database administrator and is not yet available through this
interface.</p>
<p>name=Class:Landmark:start..end</p>
<p>As above, but restricted to the designated range.</p>
<p>If you use multiple name options, then this script will generate an
overview image showing the position of each landmark.  The alias &quot;q&quot;
can be used to shorten the length of the URL.</p>
</dd>
<dt><strong><a name="type" class="item">type (Alias: t)</a></strong></dt>

<dd>
<p>This argument lists the feature types to display. The value of this
argument is a list of track names separated by spaces (&quot;+&quot; characters
when URL-escaped). For example:</p>
<pre>
      &lt;img src=&quot;<a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=mec-3</a>;
                   type=tRNA+NG+WABA+CG+ESTB&quot;&gt;</pre>
<p>Multiple type= arguments will be combined to form a single
space-delimited list. The alias &quot;t&quot; can be used to shorten the length
of the URL.</p>
<p>If the track name has a space in it, put quotes around the name:</p>
<pre>
          type=&quot;microbe tRNA&quot;+NG+WABA+CG+ESTB</pre>
</dd>
<dt><strong><a name="width" class="item">width (Alias: w)</a></strong></dt>

<dd>
<p>Width of the desired image, in pixels.</p>
</dd>
<dt><strong><a name="options" class="item">options (Alias: o)</a></strong></dt>

<dd>
<p>A space-delimited list (&quot;+&quot; characters when URL-escaped) of
mnemonic/option pairs describing how features should be formatted.
Options are integers from 0 to 3, where 0=auto, 1=compact, 2=expanded,
3=expanded and labeled. For example, to specify that the tRNA and NG
tracks should always be expanded and labeled, but that the WABA track
should be compact, use:</p>
<pre>
      options=tRNA+3+NG+3+WABA+1</pre>
<p>The alias &quot;o&quot; can be used to shorten the length of the URL.</p>
</dd>
<dt><strong><a name="abs" class="item">abs</a></strong></dt>

<dd>
<p>Use absolute (chromosome) coordinates when displaying the image.  An
argument of abs=1 will turn on absolute coordinate addressing.  An
argument of abs=0 will use relative coordinate addressing (this is the
default).</p>
</dd>
<dt><strong><a name="add" class="item">add (Alias: a)</a></strong></dt>

<dd>
<p>Superimpose one or more additional features on top of the view.
Features are specified as space (&quot;+&quot;) delimited lists in the following
format:</p>
<pre>
      add=Landmark+Type+Name+start..end,start..end,start..end</pre>
<p>&quot;Landmark&quot; is the landmark name, &quot;Type&quot; is a descriptive type that
will be printed in the image caption, &quot;Name&quot; is a name for the feature
to be printed above it, and start..end is a comma-delimited list of
ranges for discontinuous feature. Names that contain white space must
be quoted, for example &quot;BLAST hit&quot;.  Note that this all has to be
URL-escaped, so an additional feature named &quot;Your Sequence&quot;, type
&quot;Blast Hit&quot;, that is located on chromosome III in a gapped range
between 20000 and 22000, will be formatted as:</p>
<pre>
   add=III+%22Blast%20Hit%22+%22Your%20Sequence%22+20000..21000,21550..22000</pre>
<p>One or both of the type and name can be omitted. If omitted, type will
default to &quot;Your Features&quot; and the name will default to &quot;Feature XX&quot;
where XX is an integer. This allows for a very simple feature line:</p>
<pre>
      add=III+20000..21000,21550..22000</pre>
<p>Multiple add= arguments are allowed. The alias &quot;a&quot; can be used to
shorten the length of the URL.</p>
</dd>
<dt><strong><a name="style" class="item">style</a></strong></dt>

<dd>
<p>The style argument can be used to control the rendering of additional
features added with &quot;add&quot;. It is a flattened version of the style
configuration sections described in [1]this document For example, if
you have added a &quot;Blast Hit&quot; annotation, then you can tell the
renderer to use a red arrow for this glyph in this way:
style=%22Blast%20Hit%22+glyph=arrow+fgcolor=red</p>
</dd>
<dt><strong><a name="keystyle" class="item">keystyle (Alias: k)</a></strong></dt>

<dd>
<p>Controls the positioning of the track key. One of &quot;right&quot;, &quot;left&quot;,
&quot;between&quot; (default) or &quot;bottom&quot;</p>
</dd>
<dt><strong><a name="overview" class="item">overview</a></strong></dt>

<dd>
<p>Ordinarily the image will show the detail panel if the query region
corresponds to a single region, and the overview panel if multiple
regions match (or if a region that is too large to show
matches). Setting overview=1 will force the overview to be shown in
all cases.</p>
</dd>
<dt><strong><a name="flip" class="item">flip (Alias: f)</a></strong></dt>

<dd>
<p>Flip the image left to right.  Arguments are 0=don't flip (default),
and 1=flip.</p>
</dd>
<dt><strong><a name="embed" class="item">embed</a></strong></dt>

<dd>
<p>Generate image and a corresponding HTML imagemap in a form suitable
for embedding into a frame.</p>
</dd>
<dt><strong><a name="highlight" class="item">highlight (Alias: h)</a></strong></dt>

<dd>
<p>Highlight (in yellow) all features whose names match the indicated
value.</p>
</dd>
<dt><strong><a name="format" class="item">format</a></strong></dt>

<dd>
<p>Specify the format for the image file.  Either &quot;GD&quot; (the default) or
&quot;GD::SVG&quot; for scaleable vector graphics.</p>
</dd>
<dt><strong><a name="list" class="item">list</a></strong></dt>

<dd>
<p>If this argument is present, it will cause the script to dump out
various types of information in plain text form. Currently the two
values for this argument are sources, to dump out the list of data
sources, and types, to dump out the list of configured types.  For
list=sources, the script will return a simple text list of the data
source names. For list=types, the script will return a three-column
tab-delimited list giving the track names and feature types
corresponding to the currently-selected data source. The format is as
follows:</p>
<pre>
  Mnemonic &lt;tab&gt; Full description of feature &lt;tab&gt; [default]</pre>
<p>The third column contains the word &quot;default&quot; if the track will be
shown by default when no type argument is provided.</p>
</dd>
<dt><strong><a name="source" class="item">source</a></strong></dt>

<dd>
<p>This argument specifies the database for the images. The list of
sources can be found using list=sources.</p>
</dd>
</dl>
<p>Putting it all together, here's a working (very long) URL:</p>
<pre>
 <a href="http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=B0001;add=B0001+pcr">http://www.wormbase.org/db/seq/gbrowse_img/wormbase?name=B0001;add=B0001+pcr</a>
 +pcr1+20000..333000;add=B0001+%22cool%20knockout%22+kn2+30000..20000,10000..500
 0;type=add+CG+WTP;style=pcr+glyph=primers;style=%22cool%20knockout%22+glyph=tra
 nscript2+bgcolor=orange;abs=1</pre>
<p>If you wish to associate the image with an imagemap so that clicking
on a feature takes the user to the destination configured in the
gbrowse config file, you may do so by placing the URL in an &lt;iframe&gt;
section and using the embed=1 flag:</p>
<p>&lt;iframe
  src=&quot;http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001;embed=1&quot;
  width=&quot;100%&quot; height=&quot;250&quot;&gt; 
&lt;img
  src=&quot;http://localhost/cgi-bin/gbrowse_img/wormbase?name=B0001&quot;/&gt;
&lt;/iframe&gt;</p>
<p>Placing an &lt;img&gt; tag inside the &lt;iframe&gt; tag arranges for older
browsers that don't know about iframes to display the static image
instead.  You may need to adjust the width and height attributes in
order to avoid browsers placing scrollbars around the frame.</p>
<p>
</p>
<hr />
<h1><a name="known_bugs">KNOWN BUGS</a></h1>
<p>The configuration options for plugins are not correctly stored, so
tracks generated by annotation plugins, such as the Restriction site
annotator, will not display correctly when the image URL is generated
on one machine and then viewed on another. Uploaded files will
transfer correctly, however.</p>
<p>
</p>
<hr />
<h1><a name="support">SUPPORT</a></h1>
<p>Please address support requests to <a href="mailto:gmod-gbrowse@lists.sourceforge.net.">gmod-gbrowse@lists.sourceforge.net.</a></p>
<p>
</p>
<hr />
<h1><a name="author">AUTHOR</a></h1>
<p>Lincoln Stein <a href="mailto:lstein@cshl.org">lstein@cshl.org</a></p>
<p>Copyright (c) 2002 Cold Spring Harbor Laboratory</p>
<p>This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.</p>
<p>For additional help, see The GMOD Project pages.</p>

</body>

</html>
