<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<!-- $Id: 2003-02.html,v 1.1.1.1 2010-01-25 15:47:18 tharris Exp $-->
<html>
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">  
    <link rel="stylesheet" href="/stylesheets/wormbase.css"> 
    <title>WormBase Newsletter #2, February 2003</title>
  </head>
  <body>
    <h2>WormBase Newsletter #2, February 2003</h2>
    
    <h3>WormBase User's Survey</h3>
    <p>
      <img width=148 height=108 src="images/image001.jpg" align=left hspace=9>
      In preparation for our annual Advisory Board Meeting, we are
      conducting an on-line survey. Please access it from the WormBase
      home page. Your opinion in setting priorities and feedback on
      our quality of service would be of great help in planning the
      next year’s activities.  The survey is anonymous, but if you
      include your email, you will have a chance to win a WormBase mug
      with Thomas Boulin’s award-winning WormBase Logo.
    </p>

    <h3>Release Schedule</h3>
    <p>    
      The release schedule is now every two weeks. There have been
      occasional delays because of problems in the complex process of
      building the database caused by changes to its structure. 
    </p>

    <h3>Proteome/Incyte</h3>
    <p class=MsoNormal style='text-align:justify'>

      As many of you know, since the summer, access to
      Incyte/Proteome's useful wormPD database is now by subscription
      only.  Over the past few years, we had purposely avoided
      focusing on information already being provided by wormPD.  We
      have reset our priorities to provide users with the information
      most desired, based upon feedback at regional meetings this past
      summer.  We have obtained additional resources from the NHGRI
      for this year to provide more information as soon as possible.
      Providing a brief description of each gene which is seen at the
      top of the newly reorganized Gene page is currently a high
      priority task we have undertaken.  Curating RNAi data from
      individual papers as well as large-scale screens, and adding
      microarray data are other top priorities.
    </p>
    
    <h3>Expression patterns: Request for data and corrections</h3>
    <p>
      We are up-to-date with extracting gene expression patterns from
      published papers.  We will do our best to keep up with this
      information as new papers come out.
    </p>

    <p>
      Now that WormBase has a very useful collection of expression
      patterns, we ask for help in three ways (in order of
      priority):
    </p>
    <p>
      First, please check WormBase for your expression data. We
      apologize for any omissions or mistakes that we have made, but
      please tell us if you notice any!
    </p>

    <p>
      Second, please submit expression pattern data.  If you have old
      data that is in danger of being forgotten, submit it to WormBase
      via the homepage, click on Submit data; click on
      GeneExpression/Regulation or go directly to
      <a href="http://minerva.caltech.edu/~azurebrd/cgi-bin/expr_pattern.cgi">http://minerva.caltech.edu/~azurebrd/cgi-bin/expr_pattern.cgi</a>.
      If you are willing to share data before publication, please
      submit it and it will be attributed to you.
    </p>

    <p>
      Third, if you have a paper about to come out, feel free to send
      us a pre-print and we can get a head start on extracting the
      information; text files are useful since we can scan them for
      information about transgenes, etc.
    </p>

    <h3>A complete <i>C. elegans</i> Genome Sequence!</h3>

    <p>
      The WS90 release of WormBase made available today has a complete
      <i>C. elegans</i> genome sequence. The 100,258,171 base pairs
      are present in six contiguous segments corresponding to the six
      <i>C. elegans</i> nuclear chromosomes. On behalf of the <i>C.
      elegans</i> community we thank the <i>C.  elegans</i> Genome
      Sequencing Consortium at the Wellcome Trust Sanger Institute and
      Washington University's Genome Sequencing Center for this
      impressive accomplishment. In spite of the high quality of the
      sequence, there are undoubtedly remaining errors in the genome
      sequence, and thus the Sequencing Consortium will remain active
      in responding to and resolving any potential errors that
      arise. The Sequencing Consortium wants to know of any potential
      problems in the now contiguous six <i>C. elegans</i>
      chromosomes, for example, any evidence of deleted YACs, etc.
      Please inform Alan Coulson (alan@sanger.ac.uk) or WormBase.
    </p>

    <h3>Genomic Clone Requests</h3>
    <p>
      After January 1, 2003, please send all requests for
      <i>C. elegans</i> genomic clones
      to Audrey Fraser (aef@sanger.ac.uk) with a cc to Ratna Shownkeen
      (rs2@sanger.ac.uk); please include a FedEx or DHL account
      number.
    </p>

    <h3>Genome browser improvements </h3>
    <p>
      The genome browser now puts the track labels in the main image
      rather than in a key at the bottom.
    </p>

    <h3>The Textpresso search engine</h3>
    <p>
      WormBase has developed a search engine for
      <i>C. elegans</i> literature
      that allows queries with one or more key words as well as
      categories such as allele, gene, nucleic acid, pathway, etc.
      The search engine can be reached from the home page or directly
      at www.textpresso.org.  Individual sentences in almost 2000 full
      text papers are included; more will be added in the coming
      months.
    </p>

    <h3>Persons and Authors</h3>
    <p>
      As part of an effort to better organize and correct mistakes in
      the connections between people and their papers, WormBase
      distinguishes Persons from Authors: Persons are unique
      individuals with contact information.  We will be asking your
      help in making accurate connections between you and your
      publications.  An extremely useful practice would be to use full
      names whenever possible in papers and abstracts.  To check or
      update your contact information, please go to:
      <a href="http://minerva.caltech.edu/~azurebrd/cgi-bin/forms/author.cgi">http://minerva.caltech.edu/~azurebrd/cgi-bin/forms/author.cgi</a>.
    </p>

    <h3>RNAi</h3>
    <p>
      Not surprisingly there is a remarkable increase in RNAi
      experiments in papers.  We appreciate that these often have more
      thorough phenotypic analysis than the large scale screens.  For
      technical reasons, the large-scale data gets into WormBase
      faster even if published later.  We are working to catch up with
      the considerable backlog of the individual RNAi experiments from
      papers.  If any omission bothers you, please let us know and we
      will get it in as fast as possible.
    </p>

    <h3>RNAi search improvements</h3>
    <p>
      The RNAi search now allows you to combine phenotypes with either
      AND or OR criteria.
    </p>

    <h3>Coming Soon</h3>
    <h4>More microarray data</h4> 
    <p>
      A pipeline of data entry for Kim lab/Stanford microarrays has been
      established, and we will add data from published papers over the
      next few months.</p>
    
    <h3>People Of WormBase</h3>

    <p><b>New Curators/Progammers at the Sanger Institute</b>: Anthony
      Rogers and Chao-Kung Chen have joined the WormBase group.
      Chao-Kung is half-time with the CGC, and is the key liason between
      WormBase and the CGC.
    </p>

    <p><b>New Curators at Caltech:</b> Andrei Petcherski (Ph.D. at the
    University of Wisconsin, Madison with Judith Kimble), Carol
    Bastiani (postdoc with Melvin Simon and Paul Sternberg at Caltech)
    and Igor Anteshechkin (former postdoc with Min Han at the
    University of Colorado, Boulder) have joined WormBase as
    biological curators. Ranjana Kishore (former postdoc with Meera
    Sundaram at the University of Pennsylvania) has joined WormBase as
    a Gene Ontology (www.geneontology.org) curator.
    </p>

    <h3>WormBase Advisory Board</h3>
    <p>
    Our advisory board helps set priorities and goals (and generally
    acts like a thesis committee).  They ensure that the user
    community is best served, and that WormBase cooperates with other
    Model Organism Databases.
    </p>

    <p>Thomas Blumenthal, <i>C. elegans</i> molecular biologist</p>

    <p>Martin Chalfie, <i>C. elegans</i> developmental and neuro
    geneticist</p>

    <p>Jonathan Hodgkin, <i>C. elegans</i> geneticist; curator of the
    CGC genetic map</p>

    <p>Leon Avery, <i>C. elegans</i> neurobiologist; CGC website</p>

    <p>William Gelbart, PI FlyBase</p>

    <p>Janan Eppig, PI Mouse Genome Database (MGD)</p>

    <p>Michael Cherry, PI Saccharomyces Genome Database (SGD)</p>

    <p>Stanley Letovsky, bioinformatician.</p>

    <h3>Funding Sources</h3>
    <p>
      WormBase is in its third year of funding from the National Human
      Genome Research Institute (# P41 HG02223), and
      receives additional support from the
      British Medical Research Council.  Applications to these agencies
      are pending.</p>
    <hr>
    
  </body>
</html>
