<html>
<head>
<title>About the BLAT Algorithm</title>
</head>
<body>

<h1>About the BLAT algorithm</h1>

The following description of BLAT is from the <a href="http://genome.ucsc.edu/cgi-bin/hgBlat?hgsid=107033328">UCSC Genome Browser</a>.

<div style="margin:10px 100px 0px 100px; font-style:italic">
<P>BLAT on DNA is designed to
quickly find sequences of 95% and greater similarity of length 25 bases or
more.  It may miss more divergent or shorter sequence alignments.  It will find
perfect sequence matches of 25 bases, and sometimes find them down to 20 bases.
BLAT on proteins finds sequences of 80% and greater similarity of length 20 amino
acids or more.  In practice DNA BLAT works well on primates, and protein
blat on land vertebrates.
</P><P>BLAT is not BLAST.  DNA BLAT works by keeping an index of the entire genome
in memory.  The index consists of all non-overlapping 11-mers except for
those heavily involved in repeats.  The index takes up a bit less than
a gigabyte of RAM.  The genome itself is not kept in memory, allowing
BLAT to deliver high performance on a reasonably priced Linux box.
The index is used to find areas of probable homology, which are then
loaded into memory for a detailed alignment. Protein BLAT works in a similar
manner, except with 4-mers rather than 11-mers.  The protein index takes a little
more than 2 gigabytes.</P>
<P>BLAT was written by <A HREF="mailto:kent@soe.ucsc.edu">Jim Kent</A>.
Like most of Jim's software, interactive use on this web server is free to all.
Sources and executables to run batch jobs on your own server are available free
for academic, personal, and non-profit purposes.  Non-exclusive commercial
licenses are also available. See the 
<A HREF="http://www.kentinformatics.com" TARGET=_blank>Kent Informatics</A>
website for details.</P>
</div>

</body>
</html>

